engine/src/main/scala/cromwell/engine/workflow/MaterializeWorkflowDescriptorActor.scala

```diff
@@ -0,0 +1,165 @@
+    } yield value) getOrElse default
+  }
+
+  def workflowLogOptions(conf: Config): Option[WorkflowLogOptions] = {
+    for {
+      workflowConfig <- conf.getConfigOption("workflow-options")
+      dir <- workflowConfig.getStringOption("workflow-log-dir") if !dir.isEmpty
+      temporary <- workflowConfig.getBooleanOption("workflow-log-temporary") orElse Option(true)
+    } yield WorkflowLogOptions(Paths.get(dir), temporary)
+  }
+}
+
+class MaterializeWorkflowDescriptorActor() extends Actor with LazyLogging {
+
+  import MaterializeWorkflowDescriptorActor._
+
+  override def receive = {
+    case MaterializeWorkflow(id, workflowSourceFiles, conf) =>
+      val backend = CromwellBackend.getBackendFromOptions(workflowSourceFiles.workflowOptionsJson)
+      buildWorkflowDescriptor(id, workflowSourceFiles, backend, conf) match {
+        case scalaz.Success(descriptor) => sender() ! MaterializationSuccess(descriptor)
+        case scalaz.Failure(error) => sender() ! MaterializationFailure(
+          new IllegalArgumentException() with ThrowableWithErrors {
+            val message = s"Workflow input processing failed."
+            val errors = error
+          })
+      }
+    case unknownMsg => logger.error(s"${this.getClass.getName} received an unknown message: $unknownMsg")
+  }
+
+  private def buildWorkflowDescriptor(id: WorkflowId,
+                                      sourceFiles: WorkflowSourceFiles,
+                                      backend: Backend,
+                                      conf: Config): ErrorOr[WorkflowDescriptor] = {
+    val namespaceValidation = validateNamespace(sourceFiles.wdlSource)
+    val workflowOptionsValidation = validateWorkflowOptions(backend, sourceFiles.workflowOptionsJson)
+    (namespaceValidation |@| workflowOptionsValidation) {
+      (_, _)
+    } flatMap { case (namespace, workflowOptions) =>
+      val runtimeAttributes = validateRuntimeAttributes(namespace, backend.backendType)
+      val rawInputsValidation = validateRawInputs(sourceFiles.inputsJson)
+      (runtimeAttributes |@| rawInputsValidation) {
+        (_, _)
+      } flatMap { case (_, rawInputs) =>
+        validateCoercedInputs(rawInputs, namespace) flatMap { coercedInputs =>
+          val workflowRootPath = backend.buildWorkflowRootPath(backend.rootPath(workflowOptions), namespace.workflow.unqualifiedName, id)
+          val wfContext = new WorkflowContext(workflowRootPath)
+          val fileSystems = backend.fileSystems(workflowOptions)
+          val engineFunctions = backend.engineFunctions(fileSystems, wfContext)
+          validateDeclarations(namespace, workflowOptions, coercedInputs, engineFunctions) flatMap { declarations =>
+            WorkflowDescriptor(id, sourceFiles, workflowOptions, workflowLogOptions(conf), rawInputs, namespace, coercedInputs, declarations, backend,
+              configCallCaching(conf), lookupDockerHash(conf), wfContext, fileSystems).successNel
+          }
```

Comment:
This is an almighty triple-nesting of long anonymous functions. Split them out into `def`s and give them names